play audio in-line, use AwesomeFont, push demo to gh pages

enw opened the pull request:
v0.1.9
- play audio in-line in supported browsers
- use AwesomeFont "play" button for audio links in other browsers
- added tests for Wav, Ogg files

enw committed:
add FontAwesome play icon

enw committed:
don't track webgen-generated folders

enw committed:
play audio inline, show '#' for links w/o text

enw committed:
specify that links are to audio/mp3 files

enw committed:
add different mimetype for audio files

enw committed:
list the correct mimetype (not file extension!)

enw committed:
don't ignore markdown folder in git

enw committed:
test ogg & wav files.  (tested w/lynx and Chrome)

enw committed:
Create github_actions.yml

enw committed:
add .cargo/bin to path, run webgenr

enw committed:
test workflow to upload to github spaces

enw committed:
test quick fix - remove github-ref

ultrasaurus commented on src/document.rs:
```diff
@@ -195,13 +195,19 @@ impl Document {
                             } else {
                                 // no text event, just Event::End
-                                Cow::Borrowed("").into()
+                                Cow::Borrowed("#").into()
```
why this?

enw replied:
The intent here was to have *something* that can be clicked on if there is no label.  I picked a "#" for no good reason.  :)

Come to think of it... this is where the FontAwesome "▶️" should be!  Thoughts on that?

enw replied:
I'd probably make the FontAwesome change in another commit.  The current fix resolves the case where there is no text.  

TBH I should have split this into two PRs - one for FontAwesome and one for playing audio in-line.  Next time.  :)

ultrasaurus replied:
I assumed CSS with padding and maybe a circle around the triangle (which can be done with pure css or FontAwesome has a circle-play glyph

enw replied:
Aaah, I made this change before I did the in-line media player.  I'm not totally sure how to test this now because all modern browsers support in-line audio players.  Maybe test when we get generation of `.epub` working...

ultrasaurus commented:
running locally, I get this error in the browser console:
```
          GET http://localhost:8000/favicon.ico 404 (Not Found)
```

puzzled since I see no reference to a favicon and also cannot repro in main

ultrasaurus commented on Cargo.toml:
```diff
@@ -1,8 +1,8 @@
 [package]
 name = "webgenr"
-version = "0.1.8"  # copy non-hbs files in templates (recursively)
+version = "0.1.9" 
```
not a show stopper, but I find it helpful (for now at least) to include a comment about what is significant about this version

enw replied:
good call.  

```suggestion
version = "0.1.9" # play audio in-line, add FontAwesome ▶️ for audio
```

enw replied:
testing locally just in case Cargo chokes on the ▶️...

ultrasaurus commented on src/util.rs:
```diff
@@ -12,3 +12,44 @@ pub fn is_audio_file(url: &CowStr) -> bool {
 }
+
+// return the extension of an url as a string
+pub fn get_ext(url: &CowStr) -> String {
```
I think ideally this would avoid allocating a String, since we only use the extension to generate a mimetype, though ok to check in as is and iterate later :)

enw replied:
cool.  This is my first Rust project, and I'd like to learn the most Rustic way to do things.  :)

ultrasaurus commented on src/util.rs:
```diff
@@ -12,3 +12,44 @@ pub fn is_audio_file(url: &CowStr) -> bool {
+}
+
+// return mimetype given an extension
+pub fn get_mimetype(ext: &str) -> String {
```
likewise, I think this should return a &str or Cow rather than allocating a String that might not be needed by the caller

enw replied:
I'm learning about Rust memory usage.  This is allocated on the heap.  But.. doesn't "ownership" and "borrowing" take care of this?  e.g. when the caller goes out-of-scope, the value will be dropped.  [ownership & borrowing in rust-lang.org](https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html)

ultrasaurus replied:
it's not about memory leak, but rather best practice is to avoid allocations for performance, but I also believe in doing benchmarks and letting performance optimizations come from actual slow spots

enw commented:
> running locally, I get this error in the browser console:
> 
> ```
>           GET http://localhost:8000/favicon.ico 404 (Not Found)
> ```
> 
> puzzled since I see no reference to a favicon and also cannot repro in main

I'm unable to reproduce this. Seems like a browser thing.  What browser are you using and how are you serving this?  I have been testing locally w/`Version 110.0.5481.177 (Official Build) (arm64)` and lynx and testing using Live Server started within VSCode.

enw committed:
Update Cargo.toml about what is significant about this version.

enw committed:
update source path for GitHub pages artifact

enw committed:
use upload-pages-artifact to deploy example

enw committed:
debug stuff for Actions

test different path, add debug info to figure out where this code is running

enw committed:
do everything in one flow

having things in different (logical) categories may have been the problem with losing state between groups of steps.

enw committed:
move `environment:` to a `job` in YAML

ultrasaurus commented:
I'm testing it using an npm package https://github.com/lwsjs/local-web-server
```
ws -d _website
```

ultrasaurus commented on .gitignore:
```diff
@@ -1,6 +1,9 @@
+# files and folder generated by webgenr
 _website/
+templates
```
we need templates directory at the root, since those files are what are accessed by RustEmbed and used to auto-populate a new project

enw replied:
good catch.  We might want to add another .gitignore to the `examples` folder.

enw replied:
removing `templates` from `.gitignore`

enw committed:
don't `.gitignore` the templates folder

enw commented on .gitignore:
```diff
@@ -1,6 +1,5 @@
 # files and folder generated by webgenr
-_website/
-templates
+_website/ d
```
typo!  this should just be `_website/` with no ` d`

enw committed:
fix type-o

enw commented:
This should really be multiple commits - one to add in-line playing of audio, one to add FontAwesome, one to add CI/CD pushing of demos to GitHub pages.  @ultrasaurus - Your call as to how important git cleanliness is at this early stage in the development process!
